File: 1416-restore-the-array/1416-restore-the-array.py
class Solution:
    def numberOfArrays(self, s: str, k: int) -> int:
        MOD = 10**9 + 7
        n = len(s)
        
        @lru_cache(None)
        def dp(i):
            if i == n:
                return 1
            
            if s[i] == '0':
                return 0
            
            ans = 0
            num = 0
            
            for j in range(i, n):
                num = num*10 + (ord(s[j]) - ord('0'))
                if num > k:
                    break
                ans = (ans + dp(j + 1)) % MOD
                
            return ans
        
        return dp(0)

File: 1416-restore-the-array/1416-restore-the-array.cpp
class Solution {
public:
    int numberOfArrays(string s, int k) {
        const int MOD = 1e9 + 7;
        int n = s.size();
        int dp[n + 1];
        memset(dp, 0, sizeof(dp));
        dp[n] = 1;
        
        for (int i = n - 1; i >= 0; --i) {
            long long num = 0;
            if (s[i] == '0') {
                continue;
            }
            for (int j = i; j < n; ++j) {
                num = num*10 + (s[j] - '0');
                if (num > k) {
                    break;
                }
                dp[i] = (dp[i] + dp[j + 1]) % MOD;
            }
        }
        
        return dp[0];
    }
};